### DISCUSSION TITLE
WiFiClientSecure.h problem

### DISCUSSION THREAD

**Discussion by ozzcet** [esp8266/Arduino#8374] [Q&A]
I am so sorry to writing this, Please forgive me , However who change the TTL connection to BearSSL . While doing this did you check how much person effect this change ? I have used platformIO once upgraded library of esp8266 crashed. And after I have check blow link
[
https://github.com/esp8266/Arduino/blob/69f8cd693447404be8696527e94f3665d6318b8e/libraries/ESP8266WiFi/examples/HTTPSRequest/HTTPSRequest.ino](
https://github.com/esp8266/Arduino/blob/69f8cd693447404be8696527e94f3665d6318b8e/libraries/ESP8266WiFi/examples/HTTPSRequest/HTTPSRequest.ino)

There is no openly explanation on web. how could I extract cert files from webpage (previously I have used [https://www.youtube.com/watch?v=Wm1xKj4bKsY&t=781s](https://www.youtube.com/watch?v=Wm1xKj4bKsY&t=781s) 
Andreas Spiess perfectly explained )

what I have done and crashed...

downloaded openssl-0.9.8k_X64.zip from below link
[https://code.google.com/archive/p/openssl-for-windows/downloads](https://code.google.com/archive/p/openssl-for-windows/downloads)

and for example open https://www.webpage.com to extracted "Base-64 encoded x.509" type *.cer file
and open cmd as administration and write below 

C::\Users\cn\Desktop\openssl-0.9.8k_X64\bin>openssl x509 -pubkey -noout -in extractedFromWebPage.cer

C:\Users\cn\Desktop\openssl-0.9.8k_X64\bin>openssl x509 -fingerprint -in extractedFromWebPage.cer

after this I have prepared  as much totally same as  [https://github.com/esp8266/Arduino/blob/69f8cd693447404be8696527e94f3665d6318b8e/libraries/ESP8266WiFi/examples/HTTPSRequest/certs.h](https://github.com/esp8266/Arduino/blob/69f8cd693447404be8696527e94f3665d6318b8e/libraries/ESP8266WiFi/examples/HTTPSRequest/certs.h)

and I have prepared totally same [
https://github.com/esp8266/Arduino/blob/69f8cd693447404be8696527e94f3665d6318b8e/libraries/ESP8266WiFi/examples/HTTPSRequest/HTTPSRequest.ino](
https://github.com/esp8266/Arduino/blob/69f8cd693447404be8696527e94f3665d6318b8e/libraries/ESP8266WiFi/examples/HTTPSRequest/HTTPSRequest.ino)

After all things done and  one week works to solve the issue connection failed 
at below  

if (!client.connect(host, 443)) {
    Serial.println("Connection failed");
    return;
  }

Please, please , how could I get lower version which helps to use  client.setCACert_P(caCert, caCertLen);
If there is no way to get old version , please guide me to use new one.

Best regards.

**Comment by mcspr** [comment]
Certificate checks require time to be set
https://github.com/esp8266/Arduino/blob/69f8cd693447404be8696527e94f3665d6318b8e/libraries/ESP8266WiFi/examples/HTTPSRequest/HTTPSRequest.ino#L44-L45

Also, you need to tweak some build settings
https://github.com/esp8266/Arduino/blob/da6ec83b5fdbd5b02f04cf143dcf8e158a8cfd36/doc/esp8266wifi/bearssl-client-secure-class.rst#cpu-requirements

At the top of the certs.h you can see the command that generated it
https://github.com/esp8266/Arduino/blob/69f8cd693447404be8696527e94f3665d6318b8e/libraries/ESP8266WiFi/examples/HTTPSRequest/certs.h#L4-L5
The script is intended for the Core examples, but you might find it useful. Notice that the certificate you want is the CA one
https://github.com/esp8266/Arduino/blob/master/tools/cert.py
And it is working as-is, you should be able to see output from the Github API if you run it
```
Connecting to test12345
................
WiFi connected
IP address:
192.168.0.165
Waiting for NTP time sync: .
Current time: Sat Nov 13 22:07:11 2021
Connecting to api.github.com
Using certificate:
-----BEGIN CERTIFICATE-----
MIIDxTCCAq2gAwIBAgIQAqxcJmoLQJuPC3nyrkYldzANBgkqhkiG9w0BAQUFADBs
MQswCQYDVQQGEwJVUzEVMBMGA1UEChMMRGlnaUNlcnQgSW5jMRkwFwYDVQQLExB3
d3cuZGlnaWNlcnQuY29tMSswKQYDVQQDEyJEaWdpQ2VydCBIaWdoIEFzc3VyYW5j
ZSBFViBSb290IENBMB4XDTA2MTExMDAwMDAwMFoXDTMxMTExMDAwMDAwMFowbDEL
MAkGA1UEBhMCVVMxFTATBgNVBAoTDERpZ2lDZXJ0IEluYzEZMBcGA1UECxMQd3d3
LmRpZ2ljZXJ0LmNvbTErMCkGA1UEAxMiRGlnaUNlcnQgSGlnaCBBc3N1cmFuY2Ug
RVYgUm9vdCBDQTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAMbM5XPm
+9S75S0tMqbf5YE/yc0lSbZxKsPVlDRnogocsF9ppkCxxLeyj9CYpKlBWTrT3JTW
PNt0OKRKzE0lgvdKpVMSOO7zSW1xkX5jtqumX8OkhPhPYlG++MXs2ziS4wblCJEM
xChBVfvLWokVfnHoNb9Ncgk9vjo4UFt3MRuNs8ckRZqnrG0AFFoEt7oT61EKmEFB
Ik5lYYeBQVCmeVyJ3hlKV9Uu5l0cUyx+mM0aBhakaHPQNAQTXKFx01p8VdteZOE3
hzBWBOURtCmAEvF5OYiiAhF8J2a3iLd48soKqDirCmTCv2ZdlYTBoSUeh10aUAsg
EsxBu24LUTi4S8sCAwEAAaNjMGEwDgYDVR0PAQH/BAQDAgGGMA8GA1UdEwEB/wQF
MAMBAf8wHQYDVR0OBBYEFLE+w2kD+L9HAdSYJhoIAu9jZCvDMB8GA1UdIwQYMBaA
FLE+w2kD+L9HAdSYJhoIAu9jZCvDMA0GCSqGSIb3DQEBBQUAA4IBAQAcGgaX3Nec
nzyIZgYIVyHbIUf4KmeqvxgydkAQV8GK83rZEWWONfqe/EW1ntlMMUu4kehDLI6z
eM7b41N5cdblIZQB2lWHmiRk9opmzN6cN82oNLFpmyPInngiK3BD41VHMWEZ71jF
hS9OMPagMRYjyOfiZRYzy78aG6A9+MpeizGLYAiJLQwGXFK3xPkKmNEVX58Svnw2
Yzi9RKR/5CYrCsSXaQ3pjOLAEFe4yHYSkVXySGnYvCoCWw9E1CAx2/S6cCZdkGCe
vEsXCS+0yx5DaMkHJ8HSXPfqIbloEpw8nL+e/IBcm2PN7EeqJSdnoDfzAIJ9VNep
+OkuE6N36B9K
-----END CERTIFICATE-----

Requesting URL: /repos/esp8266/Arduino/commits/master/status
Request sent
Headers received
esp8266/Arduino CI has failed
Reply was:
==========
{"state":"pending","statuses":[],"sha":"bf2882d8e003c278017639fa964248fedb942454","total_count":0,"repository":{"id":32969220,"node_id":"MDEwOlJlcG9zaXRvcnkzMjk2OTIyMA==","name":"Arduino","full_name":"esp8266/Arduino","private":false,"owner":{"login":"esp8266","id":8943775,"node_id":"MDEyOk9yZ2FuaXphdGlvbjg5NDM3NzU=","avatar_url":"https://avatars.githubusercontent.com/u/8943775?v=4","gravatar_id":"","url":"https://api.github.com/users/esp8266","html_url":"https://github.com/esp8266","followers_url":"https://api.github.com/users/esp8266/followers","following_url":"https://api.github.com/users/esp8266/following{/other_user}","gists_url":"https://api.github.com/users/esp8266/gists{/gist_id}","starred_url":"https://api.github.com/users/esp8266/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/esp8266/subscriptions","organizations_url":"https://api.github.com/users/esp8266/orgs","repos_url":"https://api.github.com/users/esp8266/repos","events_url":"https://api.github.com/users/esp8266/events{/privacy}","received_events_url":"https://api.github.com/users/esp8266/received_events","type":"Organization","site_admin":false},"html_url":"https://github.com/esp8266/Arduino","description":"ESP8266 core for Arduino","fork":false,"url":"https://api.github.com/repos/esp8266/Arduino","forks_url":"https://api.github.com/repos/esp8266/Arduino/forks","keys_url":"https://api.github.com/repos/esp8266/Arduino/keys{/key_id}","collaborators_url":"https://api.github.com/repos/esp8266/Arduino/collaborators{/collaborator}","teams_url":"https://api.github.com/repos/esp8266/Arduino/teams","hooks_url":"https://api.github.com/repos/esp8266/Arduino/hooks","issue_events_url":"https://api.github.com/repos/esp8266/Arduino/issues/events{/number}","events_url":"https://api.github.com/repos/esp8266/Arduino/events","assignees_url":"https://api.github.com/repos/esp8266/Arduino/assignees{/user}","branches_url":"https://api.github.com/repos/esp8266/Arduino/branches{/branch}","tags_url":"https://api.github.com/repos/esp8266/Arduino/tags","blobs_url":"https://api.github.com/repos/esp8266/Arduino/git/blobs{/sha}","git_tags_url":"https://api.github.com/repos/esp8266/Arduino/git/tags{/sha}","git_refs_url":"https://api.github.com/repos/esp8266/Arduino/git/refs{/sha}","trees_url":"https://api.github.com/repos/esp8266/Arduino/git/trees{/sha}","statuses_url":"https://api.github.com/repos/esp8266/Arduino/statuses/{sha}","languages_url":"https://api.github.com/repos/esp8266/Arduino/languages","stargazers_url":"https://api.github.com/repos/esp8266/Arduino/stargazers","contributors_url":"https://api.github.com/repos/esp8266/Arduino/contributors","subscribers_url":"https://api.github.com/repos/esp8266/Arduino/subscribers","subscription_url":"https://api.github.com/repos/esp8266/Arduino/subscription","commits_url":"https://api.github.com/repos/esp8266/Arduino/commits{/sha}","git_commits_url":"https://api.github.com/repos/esp8266/Arduino/git/commits{/sha}","comments_url":"https://api.github.com/repos/esp8266/Arduino/comments{/number}","issue_comment_url":"https://api.github.com/repos/esp8266/Arduino/issues/comments{/number}","contents_url":"https://api.github.com/repos/esp8266/Arduino/contents/{+path}","compare_url":"https://api.github.com/repos/esp8266/Arduino/compare/{base}...{head}","merges_url":"https://api.github.com/repos/esp8266/Arduino/merges","archive_url":"https://api.github.com/repos/esp8266/Arduino/{archive_format}{/ref}","downloads_url":"https://api.github.com/repos/esp8266/Arduino/downloads","issues_url":"https://api.github.com/repos/esp8266/Arduino/issues{/number}","pulls_url":"https://api.github.com/repos/esp8266/Arduino/pulls{/number}","milestones_url":"https://api.github.com/repos/esp8266/Arduino/milestones{/number}","notifications_url":"https://api.github.com/repos/esp8266/Arduino/notifications{?since,all,participating}","labels_url":"https://api.github.com/repos/esp8266/Arduino/labels{/name}","releases_url":"https://api.github.com/repos/esp8266/Arduino/releases{/id}","deployments_url":"https://api.github.com/repos/esp8266/Arduino/deployments"},"commit_url":"https://api.github.com/repos/esp8266/Arduino/commits/bf2882d8e003c278017639fa964248fedb942454","url":"https://api.github.com/repos/esp8266/Arduino/commits/bf2882d8e003c278017639fa964248fedb942454/status"}
==========
Closing connection
```

**Reply by ozzcet** [reply]
Thank you so much, I am going to try ...
Best regards.

**Reply by ozzcet** [reply]
@mcspr thanks for your help 
After I have tried to generate cert file with cert.py file for github i have faced below problem and I cannot handle it , may you guide me please.
Best regards.
----------------------------------------------------------------------------------------------------------------------
D:\t\k\inv.\espTestPlace\HTTPS-for-Makers-master>python cert.py -s api.github.com -n github

// this file is autogenerated - any modification will be overwritten
// unused symbols will not be linked in the final binary
// generated on 2021-11-15 00:15:27
// by ['cert.py', '-s', 'api.github.com', '-n', 'github']

#pragma once

////////////////////////////////////////////////////////////
// certificate chain for api.github.com:443

const char* github_host = "api.github.com";
const uint16_t github_port = 443;

// CN: *.github.com => name: __github_com
// not valid before: 2021-03-25 00:00:00
// not valid after:  2022-03-30 23:59:59
Traceback (most recent call last):
  File "cert.py", line 125, in <module>
    sys.exit(main())
  File "cert.py", line 122, in main
    return get_certificate(server, port, args.name)
  File "cert.py", line 88, in get_certificate
    printData(ssock.getpeercert(binary_form=True))
  File "cert.py", line 49, in printData
    fingerprint = xcert.fingerprint(hashes.SHA1()).hex(':')
TypeError: hex() takes no arguments (1 given)

**Reply by mcspr** [reply]
[Looks like python version issue, you need at least Python 3.8 for `hex(":")` to work](https://docs.python.org/3/library/stdtypes.html#bytes.hex)

**Reply by ozzcet** [reply]
@mcspr  how much I could say thank you so much... It is running now. Thanks alot.... 

Best regards.
